File: lib/python/src/wg_portal_client/peers.py
from ._http import _HttpClient
from ._models import WgPeer


class PeersApi:
    """API methods for WireGuard peers."""

    def __init__(self, http: _HttpClient) -> None:
        self._http = http

    def get_by_id(self, id: str) -> WgPeer:
        return self._http.get(f"/peer/by-id/{id}")  # type: ignore[return-value]

    def get_by_interface(self, interface_id: str) -> list[WgPeer]:
        return self._http.get(f"/peer/by-interface/{interface_id}")  # type: ignore[return-value]

    def get_by_user(self, user_id: str) -> list[WgPeer]:
        return self._http.get(f"/peer/by-user/{user_id}")  # type: ignore[return-value]

    def prepare(self, interface_id: str) -> WgPeer:
        return self._http.get(f"/peer/prepare/{interface_id}")  # type: ignore[return-value]

    def create(self, data: WgPeer) -> WgPeer:
        return self._http.post("/peer/new", data)  # type: ignore[return-value]

    def update(self, id: str, data: WgPeer) -> WgPeer:
        return self._http.put(f"/peer/by-id/{id}", data)  # type: ignore[return-value]

    def delete(self, id: str) -> None:
        self._http.delete(f"/peer/by-id/{id}")

File: lib/rust/src/peers.rs
use crate::http::{HttpClient, WgPortalError};
use crate::models::WgPeer;

pub struct PeersApi<'a> {
    http: &'a HttpClient,
}

impl<'a> PeersApi<'a> {
    pub(crate) fn new(http: &'a HttpClient) -> Self {
        Self { http }
    }

    pub fn get_by_id(&self, id: &str) -> Result<WgPeer, WgPortalError> {
        self.http.get(&format!("/peer/by-id/{}", id), None)
    }

    pub fn get_by_interface(&self, interface_id: &str) -> Result<Vec<WgPeer>, WgPortalError> {
        self.http
            .get(&format!("/peer/by-interface/{}", interface_id), None)
    }

    pub fn get_by_user(&self, user_id: &str) -> Result<Vec<WgPeer>, WgPortalError> {
        self.http.get(&format!("/peer/by-user/{}", user_id), None)
    }

    pub fn prepare(&self, interface_id: &str) -> Result<WgPeer, WgPortalError> {
        self.http
            .get(&format!("/peer/prepare/{}", interface_id), None)
    }

    pub fn create(&self, data: &WgPeer) -> Result<WgPeer, WgPortalError> {
        self.http.post("/peer/new", data)
    }

    pub fn update(&self, id: &str, data: &WgPeer) -> Result<WgPeer, WgPortalError> {
        self.http.put(&format!("/peer/by-id/{}", id), data)
    }

    pub fn delete(&self, id: &str) -> Result<(), WgPortalError> {
        self.http.delete(&format!("/peer/by-id/{}", id))
    }
}
